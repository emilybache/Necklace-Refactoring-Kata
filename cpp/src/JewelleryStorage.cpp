#include <utility>
#include <sstream>
#include <iostream>
#include <vector>
#include <string>


#include "JewelleryStorage.h"

using namespace std;


string PrintList(const vector<JewelleryBase*> items)
{
    std::stringstream ss;
    for (const auto& item : items)
    {
        if (!ss.str().empty())
            ss << ", ";
        ss << *item;
    }
    return ss.str();
}

ostream &operator<<(ostream &os, const JewelleryStorage& storage) {
    os << "Jewellery Storage:\n";
    os << "Box:\n";
    os << "    Ring Compartment: " << PrintList(storage.Box.RingCompartment) << "\n";
    os << "    Top Shelf:        " << PrintList(storage.Box.TopShelf) << "\n";
    os << "    Main Section:     " << PrintList(storage.Box.MainSection) << "\n";
    os << "Tree:                 " << PrintList(storage.Tree) << "\n";
    os << "Travel Roll:          " << PrintList(storage.TravelRoll) << "\n";
    os << "Safe:                 " << PrintList(storage.Safe) << "\n";
    os << "On top of dresser:    " << PrintList(storage.DresserTop) << "\n";
    return os;
}



