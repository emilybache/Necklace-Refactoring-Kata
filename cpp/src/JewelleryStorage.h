

#ifndef NECKLACE_REFACTORING_KATA_JEWELLERYSTORAGE_H
#define NECKLACE_REFACTORING_KATA_JEWELLERYSTORAGE_H

#include <utility>
#include <string>
#include <vector>
#include <algorithm>
#include "Jewellery.h"
#include "Necklace.h"
#include "Earring.h"

using namespace std;

class JewelleryBox {
public:
    vector<JewelleryBase*> RingCompartment;
    vector<JewelleryBase*> TopShelf;
    vector<JewelleryBase*> MainSection;

    JewelleryBox() {
        // Vectors are automatically initialized
    }
};

class JewelleryStorage {
private:
    string _name;

public:
    explicit JewelleryStorage(string name): _name(std::move(name)), Box(JewelleryBox()) {};
    friend ostream& operator<<(ostream& os, const JewelleryStorage& storage);

    vector<JewelleryBase*> Safe;
    JewelleryBox Box;
    vector<JewelleryBase*> Tree;
    vector<JewelleryBase*> TravelRoll;
    vector<JewelleryBase*> DresserTop;

    JewelleryStorage() : Box(JewelleryBox()) {
        // Vectors are automatically initialized
    }

    bool IsInTravelRoll(JewelleryBase* item) {
        return std::find(TravelRoll.begin(), TravelRoll.end(), item) != TravelRoll.end();
    }
};


#endif //NECKLACE_REFACTORING_KATA_JEWELLERYSTORAGE_H
