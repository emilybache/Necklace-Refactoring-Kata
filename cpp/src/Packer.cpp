

#include "Packer.h"


void Packer::PackNecklace(Necklace* item, JewelleryStorage* storage)
{
    if (item->Stone() == Jewel::Diamond)
        storage->Safe.push_back(item);
    else if (!item->IsHeavy())
        storage->Box.TopShelf.push_back(item);
    else if (auto pendantNecklace = dynamic_cast<PendantNecklace*>(item))
    {
        storage->Tree.push_back(pendantNecklace->Chain());
        storage->Box.TopShelf.push_back(pendantNecklace->Pendant());
    }
    else
        storage->Tree.push_back(item);
}

void Packer::Pack(JewelleryBase* item, JewelleryStorage* storage)
{
    if (storage->IsInTravelRoll(item) && !item->IsHeavy())
        storage->Box.TopShelf.push_back(item);
    else if (item->Stone() == Jewel::Diamond)
        storage->Safe.push_back(item);
    else if (auto pendantNecklace = dynamic_cast<PendantNecklace*>(item))
    {
        storage->Tree.push_back(pendantNecklace->Chain());
        storage->Box.TopShelf.push_back(pendantNecklace->Pendant());
    }
    else if (item->IsSmall())
        storage->Box.TopShelf.push_back(item);
    else if (dynamic_cast<Necklace*>(item) && !item->IsHeavy()) {
        storage->Box.TopShelf.push_back(item);
    }
    else if (auto earring1 = dynamic_cast<Earring*>(item))
    {
        if (earring1->Type() == EarringType::Hoop)
            storage->Tree.push_back(earring1);
        else if (auto earring2 = dynamic_cast<Earring*>(item))
        {
            if (earring2->Type() == EarringType::Drop && earring2->Stone() != Jewel::Plain)
                storage->Box.TopShelf.push_back(earring2);
            else if (earring2->Type() == EarringType::Drop)
                storage->Box.MainSection.push_back(earring2);
        }
    }
    else if (dynamic_cast<Necklace*>(item))
        storage->Tree.push_back(item);
    else
        storage->DresserTop.push_back(item);
    
    if (storage->IsInTravelRoll(item)) {
        auto it = std::find(storage->TravelRoll.begin(), storage->TravelRoll.end(), item);
        if (it != storage->TravelRoll.end())
            storage->TravelRoll.erase(it);
    }
}
