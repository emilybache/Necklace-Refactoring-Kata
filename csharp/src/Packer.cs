using NecklaceRefactoringKata.Enums;
using NecklaceRefactoringKata.JewelleryTypes;

namespace NecklaceRefactoringKata
{
    public static class Packer
    {
        public static void PackNecklace(Necklace item, JewelleryStorage storage)
        {
            if (item.Stone == Jewel.Diamond)
                storage.Safe.Add(item);
            else if (!item.IsLarge())
                storage.Box.TopShelf.Add(item);
            else if (item is PendantNecklace pendantNecklace)
            {
                storage.Tree.Add(pendantNecklace.Chain);
                storage.Box.TopShelf.Add(pendantNecklace.Pendant);
            }
            else
                storage.Tree.Add(item);
        }

        public static void Pack(JewelleryBase item, JewelleryStorage storage)
        {
            if (storage.IsInTravelRoll(item) && !item.IsLarge())
                storage.Box.TopShelf.Add(item);
            else if (item.Stone == Jewel.Diamond)
                storage.Safe.Add(item);
            else if (item is Necklace necklace1 && necklace1 is PendantNecklace pendantNecklace)
            {
                storage.Tree.Add(pendantNecklace.Chain);
                storage.Box.TopShelf.Add(pendantNecklace.Pendant);
            }
            else if (item.IsSmall())
                storage.Box.TopShelf.Add(item);
            else if (item is Necklace && !item.IsLarge())
                storage.Box.TopShelf.Add(item);
            else if (item is Earring earring1 && earring1.Type == EarringType.Hoop)
                storage.Tree.Add(earring1);
            else if (item is Earring earring2 && earring2.Type == EarringType.Drop && earring2.Stone != Jewel.Plain)
                storage.Box.TopShelf.Add(earring2);
            else if (item is Earring earring3 && earring3.Type == EarringType.Drop)
                storage.Box.MainSection.Add(earring3);
            else if (item is Necklace necklace2)
                storage.Tree.Add(necklace2);

            else
                storage.DresserTop.Add(item);

            if (storage.IsInTravelRoll(item))
                storage.TravelRoll.Remove(item);
        }
    }
}