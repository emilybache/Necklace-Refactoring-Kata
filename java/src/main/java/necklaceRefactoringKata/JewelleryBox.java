package necklaceRefactoringKata;

import necklaceRefactoringKata.jewellery.Jewellery;
import necklaceRefactoringKata.jewellery.Ring;

import java.util.ArrayList;
import java.util.List;

public class JewelleryBox {
    public final List<Ring> ringCompartment;
    public final List<Jewellery> topShelf;
    public final List<Jewellery> mainSection;

    public JewelleryBox() {
        this.ringCompartment = new ArrayList<>();
        this.topShelf = new ArrayList<>();
        this.mainSection = new ArrayList<>();
    }
}
