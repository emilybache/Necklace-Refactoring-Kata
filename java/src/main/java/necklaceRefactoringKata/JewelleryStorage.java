package necklaceRefactoringKata;

import necklaceRefactoringKata.jewellery.Jewellery;

import java.util.ArrayList;
import java.util.List;

public class JewelleryStorage {

    public final JewelleryBox box;
    public final List<Jewellery> tree;
    public final List<Jewellery> travelRoll;
    public final List<Jewellery> safe;
    public final List<Jewellery> dresserTop;

    public JewelleryStorage() {
        this.box = new JewelleryBox();
        this.tree = new ArrayList<>();
        this.travelRoll = new ArrayList<>();
        this.safe = new ArrayList<>();
        this.dresserTop = new ArrayList<>();
    }

    public boolean IsInTravelRoll(Jewellery item) {
        return this.travelRoll.contains(item);
    }
}
