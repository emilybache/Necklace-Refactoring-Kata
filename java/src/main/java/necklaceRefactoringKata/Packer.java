package necklaceRefactoringKata;

import necklaceRefactoringKata.jewellery.Earring;
import necklaceRefactoringKata.jewellery.EarringType;
import necklaceRefactoringKata.jewellery.Jewel;
import necklaceRefactoringKata.jewellery.Jewellery;
import necklaceRefactoringKata.jewellery.Necklace;
import necklaceRefactoringKata.jewellery.PendantNecklace;

public class Packer {

    public static void packNecklace(Necklace item, JewelleryStorage storage) {
        if (item.stone == Jewel.Diamond) {
            storage.safe.add(item);
        } else if (!item.isLarge()) {
            storage.box.topShelf.add(item);
        } else if (item instanceof PendantNecklace pendantNecklace) {
            storage.tree.add(pendantNecklace.chain);
            storage.box.topShelf.add(pendantNecklace.pendant);
        } else {
            storage.tree.add(item);
        }
    }

    public static void pack(Jewellery item, JewelleryStorage storage) {
        if (storage.IsInTravelRoll(item) && !item.isLarge()) {
            storage.box.topShelf.add(item);
        } else if (item.stone == Jewel.Diamond) {
            storage.safe.add(item);
        } else if (item.isSmall()) {
            storage.box.topShelf.add(item);
        } else if (item instanceof Earring earring && earring.type == EarringType.Hoop) {
            storage.tree.add(earring);
        } else if (item instanceof Earring earring && earring.type == EarringType.Drop && earring.stone != Jewel.Plain) {
            storage.box.topShelf.add(earring);
        } else if (item instanceof Earring earring && earring.type == EarringType.Drop) {
            storage.box.mainSection.add(earring);
        } else if (item instanceof PendantNecklace pendantNecklace) {
            storage.tree.add(pendantNecklace.chain);
            storage.box.topShelf.add(pendantNecklace.pendant);
        } else if (item instanceof Necklace necklace) {
            storage.tree.add(necklace);
        } else {
            storage.dresserTop.add(item);
        }

        if (storage.IsInTravelRoll(item))
            storage.travelRoll.remove(item);
    }
}
