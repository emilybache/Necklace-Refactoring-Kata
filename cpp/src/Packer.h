

#ifndef NECKLACE_REFACTORING_KATA_PACKER_H
#define NECKLACE_REFACTORING_KATA_PACKER_H

#include "JewelleryStorage.h"
#include "Necklace.h"

using namespace std;

class Packer {
public:
    static void Pack(JewelleryBase* item, JewelleryStorage* storage);
    static void PackNecklace(Necklace* item, JewelleryStorage* storage);

};


#endif //NECKLACE_REFACTORING_KATA_PACKER_H
