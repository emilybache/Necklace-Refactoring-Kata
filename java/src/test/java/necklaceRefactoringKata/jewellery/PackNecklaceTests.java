package necklaceRefactoringKata.jewellery;

import necklaceRefactoringKata.JewelleryStorage;
import necklaceRefactoringKata.Packer;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

class PackNecklaceTests {
    private JewelleryStorage storage = new JewelleryStorage();

    String PackItem(Necklace item, JewelleryStorage storage) {
        var log = "Packing item " + item;
        if (storage.IsInTravelRoll(item))
            log += " (is in travel roll)";
        Packer.packNecklace(item, storage);
        log += "\n";
        log += Printers.PrintJewelleryStorage(storage);
        return log;
    }

    @Test
    public void PackPearlNecklace() {
        var item = new Necklace(NecklaceType.Beads, Jewel.Pearl);
        Approvals.verify(PackItem(item, storage));
    }

    @Test
    public void PackAmberNecklace() {
        var item = new Necklace(NecklaceType.Beads, Jewel.Amber);
        Approvals.verify(PackItem(item, storage));
    }

    @Test
    public void PackAmberChainNecklace() {
        var item = new Necklace(NecklaceType.Chain, Jewel.Amber);
        Approvals.verify(PackItem(item, storage));
    }

    @Test
    public void PackDiamondNecklace() {
        var item = new Necklace(NecklaceType.Chain, Jewel.Diamond);
        Approvals.verify(PackItem(item, storage));
    }

    @Test
    public void PackChain() {
        var item = new Necklace(NecklaceType.Chain, Jewel.Plain);
        Approvals.verify(PackItem(item, storage));
    }

    @Test
    public void PackLongChain() {
        var item = new Necklace(NecklaceType.LongChain, Jewel.Plain);
        Approvals.verify(PackItem(item, storage));
    }


    @Test
    public void PackPendantNecklace() {
        var item = new PendantNecklace(Jewel.Amber, NecklaceType.Chain,
                new Necklace(NecklaceType.Chain, Jewel.Plain),
                new Pendant(Jewel.Amber));
        Approvals.verify(PackItem(item, storage));
    }

    @Test
    public void PackPendantNecklaceLongChain() {
        var item = new PendantNecklace(Jewel.Amber, NecklaceType.LongChain,
                new Necklace(NecklaceType.LongChain, Jewel.Plain),
                new Pendant(Jewel.Amber));
        Approvals.verify(PackItem(item, storage));
    }
}
