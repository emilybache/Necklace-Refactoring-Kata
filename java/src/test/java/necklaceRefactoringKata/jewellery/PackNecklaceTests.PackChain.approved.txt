Packing item Necklace{type=Chain, stone=Plain}
Jewellery Storage: 
Box:
    Ring Compartment: []
    Top Shelf:        [Necklace{type=Chain, stone=Plain}]
    Main Section:     []
Tree:                 []
Travel Roll:          []
Safe:                 []
On top of dresser:    []
